### PR TITLE
Add support for vim mode in Text Editor

### PR DESCRIPTION
**Pull request by krassowski**
Relates to https://github.com/jupyterlab/jupyterlab/issues/14599

### Code changes
- Splits `VimEditorManager` out of `VimCellManager`
- Adds a file-editor specific shortcut to workaround <kbd>escape</kbd> being stolen by (hidden) search box in file editor

### User facing changes
- Adds an option to enable vim in file editors too (on by default)
- Allows to close search box with <kbd>escape</kbd> when in normal vim mode; there are two sets of shortcuts one for 4.0 and one for 4.1 (which will include https://github.com/jupyterlab/jupyterlab/pull/14421)

Screenshot from 2023-07-31 19-48-14

Screenshot from 2023-07-31 19-44-53

### CODE REVIEW

**Comment by ianhi** [comment]
I'm finding that whenver I enter an editor I'm entering in insert mode, rather than normal mode (which is what we do for notebooks) is it easy to enter into normal mode for consistency?

File: src/codemirrorCommands.ts [hunk 1]
```diff
@@ -97,198 +103,235 @@ export class VimCellManager {
...
+              Vim.map(command, keys, context);
             } else {
-              lvim.noremap(command, keys, context);
+              Vim.noremap(command, keys, context);
             }
           }
         }
       );
 
       Vim.handleKey(lcm, '<Esc>');
```
Comment (ianhi) [inline src/codemirrorCommands.ts:114]: I don't see why this doesn't put the fileeditor into normal mode. It really seems as though it should be

Comment (krassowski) [reply src/codemirrorCommands.ts:114]: There is a race condition, when it first fires the file editor has not initialised yet.

**Comment by ianhi** [comment]
one other weird thing, it seems that `modifyEditor` gets called every time I hit `Esc` when in a fileeditor. Probably not optimal, but also doesn't seem to be a big deal

File: src/codemirrorCommands.ts (hunk 1 above)
Comment (krassowski) [reply src/codemirrorCommands.ts:115]: [Should ](https://github.com/jupyterlab-contrib/jupyterlab-vim/pull/90/commits/bf589e5dfc919446acfa99c55bb453e38526cc9e)

**Comment by krassowski** [comment]
> one other weird thing, it seems that modifyEditor gets called every time I hit Esc when in a fileeditor. Probably not optimal, but also doesn't seem to be a big deal

This is because of the shortcut in schema bound to `vim:enter-normal-mode` which is a workaround for search box stealing <kbd>Esc</kbd>. It should be innocent as the heavy operations in `modifyEditor` are guarded.

**Comment by krassowski** [comment]
> I'm finding that whenver I enter an editor I'm entering in insert mode, rather than normal mode (which is what we do for notebooks) is it easy to enter into normal mode for consistency?

Resolved in https://github.com/jupyterlab-contrib/jupyterlab-vim/pull/90/commits/bf589e5dfc919446acfa99c55bb453e38526cc9e

**Comment by ianhi** [comment]
perfect! Thank you again! I think this would have taken me days of confusion to get something like this going.

**Review by ianhi — APPROVED** [review]

**Comment by krassowski** [comment]
Thank you for super quick review and releases!

**Pull request merged**